### ISSUE TITLE
Compile Issue v3.4.0 on ARM - MusicBrainz5 Library

### ISSUE THREAD

**Issue by bpm1992** [ElvishArtisan/rivendell#601]
Hi, I'm running into issues compiling v3.4.0 on the ARM-64 Architecture.  

Installed Libraries:
```
libmusicbrainz5-2/bionic,now 5.1.0+git20150707-7 arm64 [installed]
libmusicbrainz5-dev/bionic,now 5.1.0+git20150707-7 arm64 [installed]
libmusicbrainz5cc2v5/bionic,now 5.1.0+git20150707-7 arm64 [installed]
```

Here is the compile error.

``` 
CXXLD    caed
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRelease::ArtistCredit() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CQuery::~CQuery()'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRecording::ISRCList() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CList::Item(int) const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CQuery::CQuery(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRecording::Title[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `typeinfo for MusicBrainz5::CRelease'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CMetadata::~CMetadata()'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CNameCredit::JoinPhrase[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRelease::Title[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `CoverArtArchive::CCoverArt::LastErrorMessage[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRelease::MediumList() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CNameCredit::Artist() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CLabelInfo::Label() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CLabel::Name[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CMetadata::Release() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `typeinfo for MusicBrainz5::CNameCredit'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CMetadata::Disc() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CTrack::Recording() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `typeinfo for MusicBrainz5::CMedium'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CMedium::Format[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CISRC::ID[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CDisc::ReleaseList() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRelease::Barcode[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CMediumList::~CMediumList()'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `typeinfo for MusicBrainz5::CLabelInfo'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CArtistCredit::NameCreditList() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRelease::ID[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CQuery::LastHTTPCode() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CMedium::TrackList() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `CoverArtArchive::CCoverArt::FetchFront(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CQuery::LastErrorMessage[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRelease::MediaMatchingDiscID(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `CoverArtArchive::CCoverArt::CCoverArt(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CArtist::Name[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CList::NumItems() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRelease::Date[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `typeinfo for MusicBrainz5::CTrack'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRelease::LabelInfoList() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CQuery::Query(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&)'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `CoverArtArchive::CCoverArt::~CCoverArt()'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CQuery::LastResult() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `MusicBrainz5::CRecording::ID[abi:cxx11]() const'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `typeinfo for MusicBrainz5::CISRC'
/usr/local/src/rivendell-3.4.0/lib/.libs/librd.so: undefined reference to `typeinfo for MusicBrainz5::CEntity'
collect2: error: ld returned 1 exit status
Makefile:484: recipe for target 'caed' failed
make[1]: *** [caed] Error 1
make[1]: Leaving directory '/usr/local/src/rivendell-3.4.0/cae'
Makefile:518: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

**Comment by ElvishArtisan** [comment]
Looks like Autoconf/pkgconfig has not located the headers for Libmusicbrainz5. correctly What do you get when running the following:

     pkg-config --cflags libmusicbrainz5
     pkg-config --libs libmusicbrainz5

Also, what does your 'config.log' say?

**Comment by bpm1992** [comment]
Thanks. I just tried a fresh build with the latest branch along with the two commands above and still get the same result.

Here are what appear to be the relevant lines from the config.log. (attached)

```
configure:16774: checking for MUSICBRAINZ
configure:16781: $PKG_CONFIG --exists --print-errors "libdiscid libmusicbrainz5 libcoverart"
configure:16784: $? = 0
configure:16798: $PKG_CONFIG --exists --print-errors "libdiscid libmusicbrainz5 libcoverart"
configure:16801: $? = 0
configure:16839: result: yes
```

Here is what is installed:
```
libmusicbrainz5-2/bionic,now 5.1.0+git20150707-7 arm64 [installed]
libmusicbrainz5-dev/bionic,now 5.1.0+git20150707-7 arm64 [installed]
libmusicbrainz5cc2v5/bionic,now 5.1.0+git20150707-7 arm64 [installed]
libcoverart-dev/bionic,now 1.0.0+git20150706-6 arm64 [installed]
libcoverart1/bionic,now 1.0.0+git20150706-6 arm64 [installed]
libdiscid-dev/bionic,now 0.6.2-2 arm64 [installed]
libdiscid0/bionic,now 0.6.2-2 arm64 [installed,automatic]
```

Here is the config.log:
[config.log](https://github.com/ElvishArtisan/rivendell/files/5006953/config.log)

**Comment by ElvishArtisan** [comment]
Hmm. We've had a number of reports [#529, #578, #583] reporting problems linking with one or more of the MusicBrainz libraries. In each case, it's turned out to be due to corrupt distro packages. This is looking like more of the same. Can you try building those dependencies yourself, and see if you still see the problem?

**Comment by bpm1992** [comment]
Thanks. I uninstalled the APT library and compiled from source, and was able to move forward.  

This is what I get with the latest version of the code.  

```
CXXLD    rdadmin
rdadmin.o: In function `MainWidget::MainWidget(RDConfig*, RDWidget*)':
/usr/local/src/rivendell/rdadmin/rdadmin.cpp:139: undefined reference to `RDCartDialog::RDCartDialog(QString*, QString*, QString*, QString const&, QWidget*)'
edit_rdlibrary.o: In function `EditRDLibrary::okData()':
/usr/local/src/rivendell/rdadmin/edit_rdlibrary.cpp:581: undefined reference to `RDLibraryConf::setCdServerType(RDLibraryConf::CdServerType) const'
/usr/local/src/rivendell/rdadmin/edit_rdlibrary.cpp:584: undefined reference to `RDLibraryConf::setMbServer(QString) const'
edit_rdlibrary.o: In function `EditRDLibrary::EditRDLibrary(RDStation*, RDStation*, QWidget*)':
/usr/local/src/rivendell/rdadmin/edit_rdlibrary.cpp:208: undefined reference to `RDLibraryConf::cdServerTypeText(RDLibraryConf::CdServerType)'
/usr/local/src/rivendell/rdadmin/edit_rdlibrary.cpp:458: undefined reference to `RDLibraryConf::cdServerType() const'
/usr/local/src/rivendell/rdadmin/edit_rdlibrary.cpp:460: undefined reference to `RDLibraryConf::mbServer() const'
edit_station.o: In function `EditStation::EditStation(QString, QWidget*)':
/usr/local/src/rivendell/rdadmin/edit_station.cpp:476: undefined reference to `RDStation::browserPath() const'
edit_station.o: In function `EditStation::okData()':
/usr/local/src/rivendell/rdadmin/edit_station.cpp:710: undefined reference to `RDStation::setBrowserPath(QString const&) const'
collect2: error: ld returned 1 exit status
Makefile:715: recipe for target 'rdadmin' failed
make[1]: *** [rdadmin] Error 1
make[1]: Leaving directory '/usr/local/src/rivendell/rdadmin'
Makefile:519: recipe for target 'all-recursive' failed
```

**Comment by ElvishArtisan** [comment]
> This is what I get with the latest version of the code.

What were the library versions, and on what distro?

**Comment by bpm1992** [comment]
I ran into the same issue with 3.4.1 on Intel Ubuntu 20.04.

Here are the library versions (all compiled from source):
libmusicbrainz-5.1.0
libdiscid-0.6.2

libcoverart does not compile on Ubuntu.  I was able to get the latest version from their GitHub and compile.  The CPP (and CC) files are exactly the same, and the functions seem to be in there too.  

```
make[1]: Entering directory '/usr/local/src/rivendell-3.4.1/cae'
  CXXLD    caed
/usr/bin/ld: /usr/local/src/rivendell-3.4.1/lib/.libs/librd.so: undefined reference to `CoverArtArchive::CCoverArt::LastErrorMessage[abi:cxx11]() const'
/usr/bin/ld: /usr/local/src/rivendell-3.4.1/lib/.libs/librd.so: undefined reference to `CoverArtArchive::CCoverArt::FetchFront(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
/usr/bin/ld: /usr/local/src/rivendell-3.4.1/lib/.libs/librd.so: undefined reference to `CoverArtArchive::CCoverArt::CCoverArt(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: /usr/local/src/rivendell-3.4.1/lib/.libs/librd.so: undefined reference to `CoverArtArchive::CCoverArt::~CCoverArt()'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:485: caed] Error 1
make[1]: Leaving directory '/usr/local/src/rivendell-3.4.1/cae'
make: *** [Makefile:518: all-recursive] Error 1
```

**Comment by danielbair** [comment]
FYI: These guys did a great job getting Rivendell 3.4.0 onto debian arm64 and amd64...
https://www.edgeradio.org.au/rivendell-on-raspberry-pi.html and they have a repo also.

**Comment by bpm1992** [comment]
Thanks.  I'm running into the same issue on the Intel platform now.  My guess is there's something amiss with the build process/libraries/environment on Ubuntu 18.04 and Ubuntu 20.04.  lib/rdmblookup.o compiles without a problem.  cae/caed causes the issues during its linking on both ARM and Intel.

**Comment by bpm1992** [comment]
On Ubuntu 20.04, I'm able to get Rivendell to compile with the repository items by passing in parameters to use the "cc" libraries.

Configure Command:
```
./configure MUSICBRAINZ_LIBS="-L/usr/local/lib -ldiscid -lmusicbrainz5cc -lcoverartcc" --disable-docbook

```

**Comment by bpm1992** [comment]
Solved for Ubuntu 18.04 on ARM and Intel.  One issue seems to be with libtool.  When a prior version of Rivendell is on the machine, libtool tries to link rdadmin with symbols in /usr/local/bin/rdlib.so.  This version (from pre-3.4) does not have the appropriate symbols in it.

Uninstall the prior version, then configure a new version to use the "CC" version of the library.  This is installed with the regular musicbrainz repos - it just seems to work.  

```
./configure MUSICBRAINZ_LIBS="-L/usr/local/lib -ldiscid -lmusicbrainz5cc -lcoverartcc" --disable-docbook
```

Unbuntu 20.04 does not work. QT4 does not have the MYSQL library.
